core/arch/arm/kernel/secure_partition.c

```diff
@@ -1181,6 +1181,31 @@ static TEE_Result handle_tpm_event_log(struct sp_ctx *ctx, void *fdt)
 	return res;
 }
 
+static TEE_Result handle_hw_features(void *fdt)
+{
+	uint32_t val __maybe_unused = 0;
+	TEE_Result res = TEE_SUCCESS;
+	int node = 0;
+
+	/*
+	 * HW feature descriptions are optional in the SP manifest, it's not an
+	 * error if we don't find any.
+	 */
+	node = fdt_node_offset_by_compatible(fdt, 0, "arm,hw-features");
+	if (node < 0)
+		return TEE_SUCCESS;
+
+	/* Modify the crc32 property only if it's already present */
+	if (!sp_dt_get_u32(fdt, node, "crc32", &val)) {
+		res = fdt_setprop_u32(fdt, node, "crc32",
+				      feat_crc32_implemented());
```

Comment:
Here you're assuming that all PEs either support or do not support CRC32, aren't you? A mixed configuration may lead to troubles otherwise. Should detection happen on boot for each core and be consolidated here instead?

Reply:
Yes, that's correct. I understand your concern and I was thinking about the same. The problem is that the SPs get initialized before the secondary cores are switched on, so the detection on the secondary cores would be too late already. It's an unfortunate situation, I'm not sure what would be the right solution.

Reply:
Why not initialize the SPs after all the cores are online?

Reply:
The SPs must be available already when control is first handed over to the NWd, since e.g. the NWd bootloader might rely on services provided by SPs. AFAIU the secondary cores are only initialized later, during Linux boot.

Reply:
> The SPs must be available already when control is first handed over to the NWd, since e.g. the NWd bootloader might rely on services provided by SPs.

OK.

> AFAIU the secondary cores are only initialized later, during Linux boot.

Correct. Then perhaps the HW capabilities in the SP could be set once for the boot CPU, then updated when all the secondary cores are up and running? The validity of this depends on how the "crc32" node is specified, and on the actual use cases of course.

Reply:
Normally the SP processes its config once when it boots up and it's not possible to change it later. I have to think about this, if we can come up with a mechanism to workaround this.

Reply:
After a bit of investigation, it looks like it should rarely or never happen in a real world system, that some cores have CRC32 and some don't. Still, we cannot rule it out completely. My proposed fix:
- when the primary core boots check if CRC is available and store the result somewhere,
- the SPs are booted with the info that's available on the primary core,
- when the secondary cores come up, check if CRC availability matches with the primary core,
- if there is a mismatch, panic.

This is rather a workaround than a solution, but at least it prevents the SPs from executing an illegal instruction in the situation when the boot core has CRC but later the SP is scheduled on a core which doesn't. All of the above could be guarded with `CFG_SECURE_PARTITION` so it doesn't affect other use cases. What do you think?

Reply:
Makes sense to me. In fact, I believe it should be a quite common approach to expect all CPUs we're executing on to be compatible with each other.

Reply:
@jenswi-linaro hopefully, yes. I agree with the proposed approach.

Reply:
Thanks for the feedback, I implemented the workaround described above. TBH the code I added looks a bit out of place, I'm wondering if it'd be better to move it to `entry_a64.S`?